src/scores/processing/block_bootstrap.py

```diff
@@ -0,0 +1,410 @@
+            'd2' is valid, but an object with only dimension 'd2' is not valid. All datasets
+            are bootstrapped according to the same random samples along available dimensions.
+        blocks: A dictionary specifying the dimension(s) to bootstrap and the block sizes to
+            use along each dimension: `{dimension: block_size}`. The keys represent the dimensions
+            to be bootstrapped, and the values indicate the block sizes along each dimension.
+            The dimension provided here should exist in the data provided as `objects`.
+        n_iteration: The number of iterations to repeat the bootstrapping process. Determines
+            how many bootstrapped arrays will be generated and stacked along the iteration
+            dimension.
+        exclude_dims: An optional parameter indicating the dimensions to be excluded during
+            bootstrapping for each object provided in `objects`. This parameter expects a list
+            of lists, where each inner list corresponds to the dimensions to be excluded for
+            the respective object. By default, the assumption is that no dimensions are
+            excluded, and all objects are bootstrapped across all specified dimensions in `blocks`.
+        circular: A boolean flag indicating whether circular block bootstrapping should be
+            performed. Circular bootstrapping means that bootstrapping continues from the beginning
+            when the end of the data is reached. By default, this parameter is set to True.
+
+    Returns:
+        If a single Dataset/DataArray (XarrayLike) is provided, the functions returns a
+        bootstrapped XarrayLike object along the "iteration" dimension. If multiple XarrayLike
+        objects are provided, it returns a tuple of bootstrapped XarrayLike objects, each stacked
+        along the "iteration" dimension.
+
+    References:
+    Wilks, Daniel S. Statistical methods in the atmospheric sciences. Vol. 100.
+      Academic press, 2011.
+    """
+    # While the most efficient method involves expanding the iteration dimension withing the
+    # universal function, this approach might generate excessively large chunks (resulting
+    # from multiplying chunk size by iterations) leading to issues with large numbers of
+    # iterations. Hence, here function loops over blocks of iterations to generate the total
+    # number of iterations.
+
+    def _max_chunk_size_mb(ds):
+        """
+        Get the max chunk size in a dataset
+        """
+
+        def size_of_chunk(chunks, itemsize):
+            """
+            Returns size of chunk in MB given dictionary of chunk sizes
+            """
+            num = 1
+            for value in chunks:
+                value = max(value) if not isinstance(value, int) else value
+                num = num * value
+            return itemsize * num / 1024**2
+
+        ds = ds.to_dataset(name="ds") if isinstance(ds, xr.DataArray) else ds
```

Comment:
I think it may be better to flip this.

`ds = ds if isinstance(ds, xr.Dataset) else ds.to_dataset(name="ds")`

that way even if `ds` is not a `DataArray` it'll be forcefully be duck-typed and will raise an error if an incorrect data type is input.